File: award-module/src/main/java/com/ninjaone/dundieawards/organization/application/dto/ActivityModel.java
package com.ninjaone.dundieawards.organization.application.dto;

import com.ninjaone.dundieawards.organization.domain.enums.ActivityStatus;
import com.ninjaone.dundieawards.organization.domain.enums.ActivityType;
import lombok.*;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;

@Builder
@Getter @Setter
@NoArgsConstructor
@AllArgsConstructor
public class ActivityModel {
    private UUID id;
    private long organizationId;
    private LocalDateTime startedAt;
    private ActivityType type;
    private ActivityStatus status;
    private Map<String, String> context;
    private String log;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    /**
     * Factory method to create a pending activity for giving an organization Dundie Awards.
     *
     * @param id        The UUID of the activity.
     * @param organizationId The ID of the organization.
     * @param context   The context of the activity.
     * @return A newly created ActivityModel instance.
     */
    public static ActivityModel pendingGiveOrganizationDundieAwards(UUID id, Long organizationId, Map<String, String> context) {
        return ActivityModel.builder()
                .id(id)
                .type(ActivityType.GIVE_ORGANIZATION_DUNDIE_AWARDS)
                .status(ActivityStatus.PENDING)
                .context(context)
                .organizationId(organizationId)
                .build();
    }

    /**
     * Marks the activity as running and sets the current timestamp as the start time.
     *
     * @return The updated ActivityModel instance.
     */
    public ActivityModel markAsRunning() {
        if (status != ActivityStatus.RUNNING) {
            this.status = ActivityStatus.RUNNING;
            this.updatedAt = LocalDateTime.now();
            this.startedAt = LocalDateTime.now();
        }
        return this;
    }

    /**
     * Update the activity status and sets the current timestamp as the update time.
     *
     * @return The updated ActivityModel instance.
     */
    public ActivityModel withStatus(ActivityStatus status) {
        if (this.status != status) {
            this.status = status;
            this.updatedAt = LocalDateTime.now();
        }
        return this;
    }
}

File: frontend-module/src/app/core/models/activity.model.ts
export interface ActivityModel {
    id: number;
    organizationId: number;
    startedAt: Date;
    type: ActivityType;
    status: ActivityStatus;
    context: { [key: string]: string };
    log: string;
    createdAt: Date;
    updatedAt: Date;
}

// Enum definitions for ActivityType and ActivityStatus
export enum ActivityType {
    GIVE_ORGANIZATION_DUNDIE_AWARDS = "GIVE_ORGANIZATION_DUNDIE_AWARDS",
    UNKNOWN = "UNKNOWN"
}

export enum ActivityStatus {
    PENDING = "PENDING",
    RUNNING = "RUNNING",
    PAUSED = "PAUSED",
    SUCCEEDED = "SUCCEEDED",
    FAILED = "FAILED",
    CANCELLED = "CANCELLED",
    UNKNOWN = "UNKNOWN"
}
